tracing: network connection tracepoints

0xB10C opened the pull request:
This adds five new tracepoints with documentation and tests for network connections:

- established connections with `net:inbound_connection` and `net:outbound_connection`
- closed connections (both closed by us or by the peer) with `net:closed_connnection`
- inbound connections that we choose to evict with `net:evicted_inbound_connection`
- connections that are misbehaving and punished with `net:misbehaving_connection`

I've been using these tracepoints for a few months now to monitor connection lifetimes, re-connection frequency by IP and netgroup, misbehavior, peer discouragement, and eviction and more. Together with the two existing P2P message tracepoints they allow for a good overview of local P2P network activity. Also sort-of addresses https://github.com/bitcoin/bitcoin/pull/22006#discussion_r636775863. 

I've been back and forth on which arguments to include. For example, `net:evicted_connection` could also include some of the eviction metrics like e.g. `last_block_time`, `min_ping_time`, ... but I've left them out for now. If wanted, this can be added here or in a follow-up. I've tried to minimize a potential performance impact by measuring executed instructions with `gdb` where possible (method described [here](https://github.com/bitcoin/bitcoin/pull/23724#issuecomment-996919963)). I don't think a few hundred extra instructions are too crucial, as connection opens/closes aren't too frequent (compared to e.g. P2P messages).   Note: e.g. `CreateNodeFromAcceptedSocket()` usually executes between 80k and 90k instructions for each new inbound connection.

| tracepoint                 	| instructions                                           	|
|----------------------------	|--------------------------------------------------------	|
| net:inbound_connection     	| 390 ins                                                	|
| net:outbound_connection    	| between 700 and 1000 ins                                     	|
| net:closed_connnection     	| 473 ins                                                	|
| net:evicted_inbound_connection     	| not measured; likely similar to net:closed_connnection 	|
| net:misbehaving_connection | not measured                                           	|

Also added a bpftrace (tested with v0.14.1) `log_p2p_connections.bt` example script that produces output similar to:
```
Attaching 6 probes...
Logging opened, closed, misbehaving, and evicted P2P connections
OUTBOUND conn to 127.0.0.1:15287: id=0, type=block-relay-only, network=0, total_out=1
INBOUND conn from 127.0.0.1:45324: id=1, type=inbound, network=0, total_in=1
MISBEHAVING conn id=1, message='getdata message size = 50001'
CLOSED conn to 127.0.0.1:15287: id=0, type=block-relay-only, network=0, established=1231006505
EVICTED conn to 127.0.0.1:45324: id=1, type=inbound, network=0, established=1612312312
```

0xB10C committed:
tracing: add inbound connection tracepoint

0xB10C committed:
tracing: add outbound connection tracepoint

0xB10C committed:
tracing: add inbound connection eviction tracepoint

0xB10C committed:
tracing: add misbehaving conn tracepoint

0xB10C committed:
tracing: connection closed tracepoint

0xB10C committed:
tracing: log_p2p_connections.bt example

A bpftrace script that logs information from the
net:*_connection tracepoints.

I've tested this script with bpftrace version 0.14.1 and v0.20.2.

jb55 commented:
nice, utACK 7252c60a7ef3b86b52aba110e2f808171d952577

russeree commented:
Just reporting in that everything is functional.

Console Log
![image](https://user-images.githubusercontent.com/3104223/185370345-804a9832-c656-451b-8395-09c16164a1fd.png)
Tests Passed
![image](https://user-images.githubusercontent.com/3104223/185369422-1d4ed3c6-b778-448c-8d23-95e25a155279.png)

0xB10C commented:
Note to self: might be worth checking if we can also pass AS information and not only the net_group.

0xB10C commented:
> Note to self: might be worth checking if we can also pass AS information and not only the net_group.

This is possible by accessing the CNodeStats, but don't think we should do that in the tracepoint. We have the IP and can lookup the AS in a tracing scirpt, if necessary. 

---

Fixed up the [linter complains](https://github.com/bitcoin/bitcoin/runs/7810747620) and rebased. This is ready for further review.

dergoegge commented:
Concept ACK

0xB10C commented:
The linter started [complaining](https://cirrus-ci.com/task/5633626411368448?logs=lint#L224) about "E275 missing whitespace after keyword". Fixed this up in the individual commits.

ac438b10918ef05df2e564734d879699c8aa18f9 -> 6de4fc73c935baa2cc6f54dd57c8e57a5df3c7ae ([2022-05-connection-tracepoints.pr.1](https://github.com/0xB10C/bitcoin/tree/2022-05-connection-tracepoints.pr.1) -> [2022-05-connection-tracepoints.pr.2](https://github.com/0xB10C/bitcoin/tree/2022-05-connection-tracepoints.pr.2); [compare](https://github.com/0xb10c/bitcoin/compare/2022-05-connection-tracepoints.pr.1..2022-05-connection-tracepoints.pr.2))

dergoegge commented on src/net.cpp:
```diff
@@ -930,6 +938,15 @@ bool CConnman::AttemptToEvictConnection()
         if (pnode->GetId() == *node_id_to_evict) {
             LogPrint(BCLog::NET, "selected %s connection for eviction peer=%d; disconnecting\n", pnode->ConnectionTypeAsString(), pnode->GetId());
+
+            TRACE6(net, evicted_connection,
+                pnode->GetId(),
+                pnode->m_addr_name.c_str(),
+                pnode->ConnectionTypeAsString().c_str(),
+                pnode->ConnectedThroughNetwork(),
+                pnode->nKeyedNetGroup,
+                std::chrono::duration<uint64_t>(pnode->m_connected).count());
```
Thinking a little bit about what data should go into the tracepoints.

`m_addr_name`: would it make more sense to put `addr` in binary form (or both)? I'm not sure they are equivalent all the time.
`nKeyedNetGroup`: is derived deterministically from `addr` using random seeds that are created with every `CConnman`. I think that means that across restarts of a node previous net groups loose meaning (i.e. you can't compare them to netgroups of new connections). Maybe `NetGroupManager::GetGroup(addr)` would be better?

virtu replied:
I can imagine some use cases where having consistent netgroup identifiers would be advantageous. Unless `NetGroupManager::GetGroup()` adds too much overhead, they seem preferable.

0xB10C replied:
Looking into `GetGroup()` I saw it returns a C++ `std::vector<unsigned char>`. We can't pass C++ `std::vector`'s in the tracepoints as the eBPF code, written in C, won't be able to read it. We can call `data()` and pass this in the tracepoint as pointer to unsigned chars (we do this with e.g. the raw P2P messages too). To read from this pointer, we also need to pass the of the length of underlying array.

I'll give this a shot in a separate branch.

0xB10C replied:
An alternative is to pass `GetMappedAS()` which is an `uint32_t` that will be !=0 in an ASMap future. A net group similar to Bitcoin Core's can be calculated from the peer address by consumers that need it.

virtu commented:
ACK [6de4fc7](https://github.com/bitcoin/bitcoin/commit/6de4fc73c935baa2cc6f54dd57c8e57a5df3c7ae)

Code reviewed tracepoints, tests, and demo script. Tracepoint placement looks good, so do tests. Successfully ran functional tests and the demo script.

Some points:
- Any reason why only `inbound_connection` and `outbound_connection` explicitly cast all arguments in the demo script?
- I think some of the format specifiers in the demo script are lightly off, leading to outputs with negative netgroup values on my system. I believe it should be  `%lld` for int64 and `%llu` for uint64.
- As [brought up previously](https://github.com/bitcoin/bitcoin/pull/25832#discussion_r1143291078), it would be nice to switch from `nKeyedNetGroup` to netgroup ids that are consistent over time

dergoegge commented on src/net.cpp:
```diff
@@ -930,6 +938,15 @@ bool CConnman::AttemptToEvictConnection()
         if (pnode->GetId() == *node_id_to_evict) {
             LogPrint(BCLog::NET, "selected %s connection for eviction peer=%d; disconnecting\n", pnode->ConnectionTypeAsString(), pnode->GetId());
+
+            TRACE6(net, evicted_connection,
```
```suggestion
            TRACE6(net, evicted_inbound_connection,
```

The logic in `net` exclusively evicts inbound peers and there is more eviction logic in `net_processing` for outbound connections. Maybe adding separate trace points for both makes sense?

0xB10C commented:
Still think this is relevant, marking this as in draft until I address the feedback.

0xB10C commented:
This is ready for review again. Changes:

- Rebased: `master` now includes https://github.com/bitcoin/bitcoin/pull/27831 and https://github.com/bitcoin/bitcoin/pull/27944. The tests in this PR have been updated to match the changes in these PRs. 
- https://github.com/bitcoin/bitcoin/pull/25832#discussion_r1143291078: I've dropped `nKeyedNetGroup` from all tracepoints and didn't add a `GetGroup()` call. The net group can be calculated from the passed address. Passing `NetGroupManager::GetGroup()` seems impractical as it's a `std::vector<unsigned char>`.
- https://github.com/bitcoin/bitcoin/pull/25832#discussion_r1152094594: Renamed `evicted_connection` to `evicted_inbound_connection` to clarify that this only covers inbound evictions. Outbound eviction is happening in multiple places in `net_processsing.cpp` and would need multiple tracepoints. If needed, these can be added in a separate pull-request.
- https://github.com/bitcoin/bitcoin/pull/25832#discussion_r1144475716: using `Ticks<std::chrono::seconds>(m_connected);` now
- adapted the bpftrace demo script and tested with bpftrace v0.18

jb55 commented:
On Mon, May 02, 2022 at 04:30:15PM +0200, 0xb10c wrote:
>---
> doc/tracing.md                        | 12 ++++
> src/net_processing.cpp                | 10 +++-
> test/functional/interface_usdt_net.py | 84 ++++++++++++++++++++++++++-
>
>diff --git a/src/net_processing.cpp b/src/net_processing.cpp
>index 6b415b3a1ea63..ca7e2c8ca3be0 100644
>--- a/src/net_processing.cpp
>+++ b/src/net_processing.cpp
>@@ -1662,13 +1662,21 @@ void PeerManagerImpl::Misbehaving(Peer& peer, int howmuch, const std::string& me
>     const std::string message_prefixed = message.empty() ? "" : (": " + message);
>     std::string warning;
>
>-    if (score_now >= DISCOURAGEMENT_THRESHOLD && score_before < DISCOURAGEMENT_THRESHOLD) {
>+    const bool discurage_threshold_exceeded = score_now >= DISCOURAGEMENT_THRESHOLD && score_before < DISCOURAGEMENT_THRESHOLD;
>+    if (discurage_threshold_exceeded) {

nit: should this be `discourage_threshold_exceeded` ? 😅

0xB10C commented:
rebased to inlcude https://github.com/bitcoin/bitcoin/pull/28629 and added the compiler flags here too.

jb55 commented:
crACK af207077578265964674618b9e2b0e40c087ae35

0xB10C commented:
Hm the CI failure of `previous releases, qt5 dev package and depends packages, DEBUG` (CI job used to find silent-merge conflicts (?)) four days ago seems weird. Didn't see any conflicts when rebasing.

Anyway, rebased. Sorry for invalidating your ACK again, @jb55. Fixed the discurage -> discourage misspelling. I thought I did it in an earlier push.

jb55 commented:
On Thu, Jan 18, 2024 at 02:03:35AM -0800, 0xB10C wrote:
>Hm the CI failure of `previous releases, qt5 dev package and depends
>packages, DEBUG` (CI job used to find silent-merge conflicts (?)) four
>days ago seems weird. Didn't see any conflicts when rebasing.
>
>Anyway, rebased. Sorry for invalidating your ACK again, @jb55. Fixed
>the discurage -> discourage misspelling. I thought I did it in an
>earlier push.

sorry I haven't been able to sit down and actually test this yet. The
changes look good though!

0xB10C commented:
Can we re-trigger the `Cirrus CI / ASan + LSan + UBSan + integer, no depends, USDT` job? The failing `wallet_send.py --descriptors` test looks unrelated.

laanwj commented on doc/tracing.md:
```diff
@@ -93,6 +93,65 @@ to user-space in full. Messages longer than a 32kb might be cut off. This can
+1. Peer ID as `int64`
+2. Peer address and port (IPv4, IPv6, Tor v3, I2P, ...) as `pointer to C-style String` (max. length 68 characters)
+3. Connection Type (inbound, feeler, outbound-full-relay, ...) as `pointer to C-style String` (max. length 20 characters)
+4. Network the peer connects from as `uint32` (1 = IPv4, 2 = IPv6, 3 = Onion, 4 = I2P, 5 = CJDNS). See `Network` enum in `netaddress.h`.
```
It's slightly inconsistent to pass this parameter as int, while the other enum (Connection Type) gets passed as string. Might similarly pass `GetNetworkName(...).c_str()`. Especially after #26593 which would remove the call overhead in the happy path.
No strong opinion though, I think the integer value is easier to use if you're going to make a histogram or such.

0xB10C replied:
I fully agree what you've said. I'm leaving as is for now. If #26593 is merged and I retouch this, I'll reconsider changing this. I'll leave this as "unresolved" for now.

fanquake replied:
I think this can be fixed in a follow up pre 29.x.

virtu reviewed (commented):
ACK [996029b](https://github.com/bitcoin/bitcoin/pull/25832/commits/996029b1fdc131b5a65c34b4898a515d523268d3) (modulo comment and [this](https://github.com/bitcoin/bitcoin/pull/25832/files/996029b1fdc131b5a65c34b4898a515d523268d3#diff-adf830d8b5ab19af24c3f9be2cb2c8a14fa49455d22613c50adfd5e3cd583013))

Successfully ran `log_p2p_connections.bt` over the weekend on both x86_64/clang and aarch64/gcc.

Extended functional test (`interface_usdt_net.py`) equally passed on both machines.

0xB10C commented on contrib/tracing/README.md:
```diff
@@ -335,4 +335,25 @@ $ python3 contrib/tracing/mempool_monitor.py ./src/bitcoind
+INBOUND conn from 127.0.0.1:45324: id=1, type=inbound, network=0, net_group=90061749176390714, total_in=1
+MISBEHAVING conn id=1, score_before=0, score_increase=20, message='getdata message size = 50001', threshold_exceeded=false
+CLOSED conn to 127.0.0.1:15287: id=0, type=block-relay-only, network=0, net_group=245974404224840066, established=1231006505
+EVICTED conn to 127.0.0.1:45324: id=1, type=inbound, network=0, net_group=90061749176390714, established=1612312312
```
dropped in 8109319e102c41d3aeed0ecfbc3a0e23b7fea807

virtu commented:
LGTM! ACK [8109319](https://github.com/bitcoin/bitcoin/commit/8109319e102c41d3aeed0ecfbc3a0e23b7fea807)

0xB10C commented:
note that https://github.com/bitcoin/bitcoin/pull/29575 removes the misbehavior score (which is exposed via the `net:misbehaving_connection` tracepoint). If/when #29575 is merged, the `misbehaving_connection` tracepoint will change.

0xB10C commented:
Must have closed this while force pushing. That was not my intention.


I've rebased this with https://github.com/bitcoin/bitcoin/pull/29575 merged. The `net:misbehaving_connection` changed from

```c++
TRACE5(net, misbehaving_connection,
  peer.m_id,
  score_before,
  howmuch,
  message.c_str(),
  discourage_threshold_exceeded);
```

to

```c++
TRACE2(net, misbehaving_connection,
  peer.m_id,
  message.c_str());
```

I've updates the tests and the bpftrace demo script accordingly.

maflcko reviewed (commented):
left two style nits, feel free to ignore

0xB10C commented:
Thanks @maflcko. I'll incorporate these suggestions once https://github.com/bitcoin/bitcoin/pull/30512 is merged and I rebase.

maflcko reviewed (approved):
Left a minor follow-up idea and a few style nits, but nothing blocking. lgtm

review ACK edef8c78e3a436e7187937b6e1bc906392ef9891 💽

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK edef8c78e3a436e7187937b6e1bc906392ef9891 💽
E1bat9/3mlxuNJ9mxUBYhUxNSOWw8eIFqjqBo86KKWEUbfBUe85vUgW0Kv+VOjkL/7yPvxAdFPVdZ/LWpOP8CA==
```

</details>

0xB10C commented on contrib/tracing/log_p2p_connections.bt:
```diff
@@ -0,0 +1,51 @@
+  printf("EVICTED conn to %s: id=%ld, type=%s, network=%d, established=%ld\n", $addr, $id, $conn_type, $network, arg4);
+}
+
+usdt:./src/bitcoind:net:misbehaving_connection
```
note to self: update the binary path to new cmake location

0xB10C commented:
rebased, addresses some of @maflcko's nits and https://github.com/bitcoin/bitcoin/pull/25832#discussion_r1741287188

i-am-yuvi commented:
tested ACK 🔥

Successfully ran demo scripts and functional tests!!

I have been using these tracepoints for 3-4 months from now and I find it very useful. They help you keep track of your peers, also I am working on a tool that detects spy peers and for that these tracepoints are very useful. 

Great work @0xB10C.

vasild commented on src/net_processing.cpp:
```diff
@@ -1945,6 +1945,10 @@ void PeerManagerImpl::Misbehaving(Peer& peer, const std::string& message)
     LogDebug(BCLog::NET, "Misbehaving: peer=%d%s\n", peer.m_id, message_prefixed);
+    TRACE2(net, misbehaving_connection,
+        peer.m_id,
+        message.c_str()
```
I am not sure how serious is violating the expectation that this is going to be max 128 chars, but if that could cause troubles, maybe here:

```suggestion
        message.substr(0, 127).c_str() // 128 including the terminating '\0'
```

0xB10C replied:
I think this can be handled on the user side. The bpftrace tool cuts of strings after 64 chars by default and bcc and libbpf both have [bpf_probe_read_user_str()](https://docs.ebpf.io/linux/helper-function/bpf_probe_read_user_str/), which would replace the last char with `\0` on overflow. 

I haven't had any problems with this using the tracepoints in Python and Rust. I did a quick check with a C program, and using `bpf_probe_read_user_str()` seems fine: https://github.com/0xB10C/libbpf-bootstrap/commit/d94496936dec11699736abce0d303a94f815f59b

vasild commented on test/functional/interface_usdt_net.py:
```diff
@@ -77,12 +112,112 @@
+int trace_misbehaving_connection(struct pt_regs *ctx) {
+    struct MisbehavingConnection misbehaving = {};
+    bpf_usdt_readarg(1, ctx, &misbehaving.id);
+    bpf_usdt_readarg_p(2, ctx, &misbehaving.message, 64);
```
64? Shouldn't that be `MAX_MISBEHAVING_MESSAGE_LENGTH` (which is 128)?

vasild commented on test/functional/interface_usdt_net.py:
```diff
@@ -165,7 +308,194 @@ def handle_outbound(_, data, __):
+        self.log.info(
+            f"connect {EXPECTED_CLOSED_CONNECTIONS} P2P test nodes to our bitcoind node")
+        testnodes = list()
+        for p2p_idx in range(EXPECTED_CLOSED_CONNECTIONS):
+            testnode = P2PInterface()
+            self.nodes[0].add_p2p_connection(testnode)
+            testnodes.append(testnode)
+        for node in testnodes:
+            node.peer_disconnect()
+        bpf.perf_buffer_poll(timeout=400)
+
+        assert_equal(EXPECTED_CLOSED_CONNECTIONS, len(closed_connections))
```
What happens if the disconnections do not happen within the timeout? Is that 400 seconds? Would it then go to the assert and fail?

Here we can deterministically wait to have 0 connected peers:

```python
self.wait_until(lambda: len(self.nodes[0].getpeerinfo()) == 0)
```

vasild replied:
Is `bpf.perf_buffer_poll(timeout=400)` needed now that there is `self.wait_until(lambda: len(self.nodes[0].getpeerinfo()) == 0)`? If not needed, then better drop it.

vasild commented on contrib/tracing/log_p2p_connections.bt:
```diff
@@ -0,0 +1,51 @@
+  $conn_type = str(arg2);
+  $network = (int32) arg3;
+  $existing = (uint64) arg4;
+  printf("INBOUND conn from %s: id=%ld, type=%s, network=%d, total=%d\n", $addr, $id, $conn_type, $network, $existing + 1);
```
I got the impression that `$existing` should be already the number of connections, including this one, why `+ 1`? (same question for the code below)

0xB10C replied:
Correct, this is an oversight from an older version of this PR. Likely related to https://github.com/bitcoin/bitcoin/pull/25832#discussion_r1692884230.

vasild reviewed (commented):
Approach ACK 5b59cfc944f9eac73e1e69bcc660280b6809724a

0xB10C commented:
Thanks for the comments @vasild. I plan to address your comments soon, when I rebase on https://github.com/bitcoin/bitcoin/pull/26593.

0xB10C commented on test/functional/interface_usdt_net.py:
```diff
@@ -77,12 +112,112 @@
+int trace_inbound_connection(struct pt_regs *ctx) {
+    struct NewConnection inbound = {};
+    bpf_usdt_readarg(1, ctx, &inbound.conn.id);
+    bpf_usdt_readarg_p(2, ctx, &inbound.conn.addr, MAX_PEER_ADDR_LENGTH);
```
> What would it do if the argument is longer than the supplied buffer? I presume the last arg is the length of the buffer, so it is not going to overflow, I guess. 

Correct, it would cut it off.

> But would it add a trailing '\0'?

No. However there is [bpf_probe_read_user_str()](https://docs.ebpf.io/linux/helper-function/bpf_probe_read_user_str/) which does that. This generally seems to be the better suited methodology here. I'll use that for the changes in this PR. There are other places in our code we can update in a potential follow up.

> That is - could it copy MAX_PEER_ADDR_LENGTH from the input to the output and leave it like that? If yes, would that be a problem - a string without a terminating '\0'?

I think, for us, in this test here, this isn't a problem. The Python ctype mapping for `Connection` defined [below](https://github.com/bitcoin/bitcoin/pull/25832/commits/3c16937f66523e92eb81afacff285fdab80e2e48#diff-e4179f6d1b7cda2cccb1b611452a520817d2d94a8c3f86f00ede8e3dd03b0b61R113-R119) has the same size for an address (`("addr", ctypes.c_char * MAX_PEER_ADDR_LENGTH)`).

0xB10C committed:
tracing: document that peer addrs can be >68 chars

A v3 onion address with a `:` and a five digit port has a length of
68 chars. As noted in
https://github.com/bitcoin/bitcoin/pull/25832#discussion_r1781040991
peers e.g. added via hostname might have a longer CNode::m_addr_name.
These might be cut off in tracing tools.

vasild commented on test/functional/interface_usdt_net.py:
```diff
@@ -77,12 +112,130 @@
+BPF_PERF_OUTPUT(inbound_connections);
+int trace_inbound_connection(struct pt_regs *ctx) {
+    struct NewConnection inbound = {};
+    uint64_t conn_type_pointer;
+    uint64_t address_pointer;
+    bpf_usdt_readarg(1, ctx, &inbound.conn.id);
+    bpf_usdt_readarg(2, ctx, &address_pointer);
+    bpf_usdt_readarg(3, ctx, &conn_type_pointer);
+    bpf_usdt_readarg(4, ctx, &inbound.conn.network);
+    bpf_usdt_readarg(5, ctx, &inbound.existing);
+    bpf_probe_read_user_str(&inbound.conn.addr, sizeof(inbound.conn.addr), (void *) address_pointer);
+    bpf_probe_read_user_str(&inbound.conn.type, sizeof(inbound.conn.type), (void *) conn_type_pointer);
```
Non blocker comment: this cast from `uint64_t` to `(void*)` looks odd, maybe it will produce compiler warning in some cases? Also, on 32 bit systems that should be `uint32_t`. Can these variables be declared as `void*` instead? Like:

```suggestion
    void* conn_type_pointer;
    void* address_pointer;
    bpf_usdt_readarg(1, ctx, &inbound.conn.id);
    bpf_usdt_readarg(2, ctx, &address_pointer);
    bpf_usdt_readarg(3, ctx, &conn_type_pointer);
    bpf_usdt_readarg(4, ctx, &inbound.conn.network);
    bpf_usdt_readarg(5, ctx, &inbound.existing);
    bpf_probe_read_user_str(&inbound.conn.addr, sizeof(inbound.conn.addr), address_pointer);
    bpf_probe_read_user_str(&inbound.conn.type, sizeof(inbound.conn.type), conn_type_pointer);
```

0xB10C replied:
Thanks! Changed the declaration from `uint64_t` to `void*`.

vasild reviewed (approved):
ACK 15dfd72f8ff47736a13c80743d5d132fd2763ecb (modulo to lint failure)

0xB10C commented:
Rebased, addressed https://github.com/bitcoin/bitcoin/pull/25832#discussion_r1905788722 and fixed the linter warning.

laanwj commented:
Both CI errors are related to mempool synchronization. It seems very unlikely to be caused by the changes in this PR.

laanwj reviewed (approved):
Code review ACK 6eb94b3082e6ff6882e27ee98abb3c79f34abbe0

0xB10C commented on test/functional/interface_usdt_net.py:
```diff
@@ -78,12 +113,130 @@
+BPF_PERF_OUTPUT(evicted_inbound_connections);
+int trace_evicted_inbound_connection(struct pt_regs *ctx) {
+    struct ClosedConnection evicted = {};
+    void* conn_type_pointer;
+    void* address_pointer;
```
nit: this and other occurences
```suggestion
    void *conn_type_pointer = NULL, *address_pointer = NULL;
```

0xB10C commented:
Rebased and addressed https://github.com/bitcoin/bitcoin/pull/25832#discussion_r1925424138

laanwj reviewed (approved):
re-ACK e3622a969293feea75cfadc8f7c6083edcd6d8de

vasild reviewed (approved):
ACK e3622a969293feea75cfadc8f7c6083edcd6d8de

sipa reviewed (commented):
utACK e3622a969293feea75cfadc8f7c6083edcd6d8de

0xB10C commented:
Using BIP155 network IDs is a good idea! Planning to address https://github.com/bitcoin/bitcoin/pull/25832#discussion_r1943088145 in a followup.

vasild commented:
`CNetAddr::GetBIP155Network()` might be useful.

0xB10C commented:
I implemented both the BIP155 NetworkID approach (https://github.com/0xB10C/bitcoin/commit/72133f356f359db818bc9eb2194fdbfa783a5d2c) and the Network as string approach from https://github.com/bitcoin/bitcoin/pull/25832#discussion_r1555936795  (https://github.com/0xB10C/bitcoin/commit/39102d4ae166e34ee4b76581c1431184c486d8d9). 

The BIP155 approach is a bit more involved with making `GetBIP155Network()` public, introducing a `ConnectedThroughBIP155Network()` helper, and passing `0` (or something else?) for `NET_INTERNAL` as there is no BIP155 equivalent. With this approach, NET_UNROUTABLE isn't exposed - e.g. `127.0.0.1` will get the BIP155 NetworkID for IPv4.

The network name as string approach exposes NET_UNROUTABLE as `not_publicly_routable` and NET_INTERNAL as `internal`. We have these in `-netinfo` too.

I have a slight preference on BIP155 NetworkIDs over keeping-as-is or network name as string.

laanwj commented:
> I have a slight preference on BIP155 NetworkIDs over keeping-as-is or network name as string.

Agree. Strings are sensitive to typos, and require extra effort for matching when doing statistics. i think even the current enumeration is better than that. BIP155 is a good suggestion, but as you have to make up non-standard values, and handle NET_UNROUTABLE differently i'm not entirely sure.

The purpose of tracing is to gain insight into internal state so logging the actual values makes more sense than thinking up a new convention. The internal enumeration is likely stable enough to warrant being a (semi-)stable interface.

0xB10C commented:
> Agree. Strings are sensitive to typos, and require extra effort for matching when doing statistics. i think even the current enumeration is better than that. BIP155 is a good suggestion, but as you have to make up non-standard values, and handle NET_UNROUTABLE differently i'm not entirely sure.
> The purpose of tracing is to gain insight into internal state so logging the actual values makes more sense than thinking up a new convention. The internal enumeration is likely stable enough to warrant being a (semi-)stable interface.

I tend to agree. A few toughs:
- Mapping the internal Network enum to other values makes it harder to go from tracepoint output/observation to Bitcoin Core internals again
- Using a third (1. `Network` enum, 2. BIP155) protocol "BIP155 extended for tracepoint interface to handle Bitcoin Core internal values not present in BIP155" seems weird, confusing, and not worth it (?). If mapping to BIP155 would have been more straight forward, I'd have been happy to use it.
- Personally, I'd favor to avoid changes/functions (like `ConnectedThroughBIP155Network()`, making `GetBIP155Network()` public) that are only used by or useful for tracepoints.
- The tracepoint interface will inherently expose internals. It won't ever be stable in a way the RPC interface is stable. If we remove or refactor something, the affected tracepoints will need to change. This happened in the past with, for example, removing the misbehaving score (https://github.com/bitcoin/bitcoin/pull/25832#issuecomment-2182585453) after https://github.com/bitcoin/bitcoin/pull/29575 and https://github.com/bitcoin/bitcoin/pull/31122#discussion_r1820876152.  If the internals change, the tracepoints and their interface needs to change too. Though, this could be better documented in https://github.com/bitcoin/bitcoin/blob/master/doc/tracing.md#semi-stable-api 

Unless there are strong opinions on needing a follow up, I'll plan to leave this as is.

0xB10C commented on test/functional/interface_usdt_net.py:
```diff
@@ -457,5 +474,43 @@ def handle_misbehaving_connection(_, data, __):
+        for closed_connection in closed_connections:
+            assert closed_connection.conn.id > 0
+            assert_equal("inbound", closed_connection.conn.conn_type.decode('utf-8'))
+            assert_equal(0, closed_connection.conn.network)
```
This should have been 
```patch
- assert_equal(0, closed_connection.conn.network)
+ assert_equal(NETWORK_TYPE_UNROUTABLE, closed_connection.conn.network)
```
